src/test/java/org/jenkinsci/plugins/workflow/test/steps/SemaphoreStep.java

```diff
@@ -206,10 +209,6 @@ public static class Execution extends AbstractStepExecutionImpl {
                 getContext().onFailure(error);
             } else {
                 LOGGER.info(() -> "Blocking " + k);
-                String c = Jenkins.XSTREAM.toXML(getContext());
-                synchronized (s) {
-                    s.contexts.put(k, c);
-                }
                 sync = false;
```

Comment:
FWIW I think there may be other pre-existing bugs here. What happens if another thread calls `success` or `failure` after the first `synchronized` block above but before this completes? Those threads will call `StepContext.onSuccess`/`StepContext.onFailure` before `StepExecution.start` has completed, which seems unusual, but maybe it's fine.

Reply:
Maybe we should change the overall approach, drop support for the synchronous completion cases, and have the step just poll continuously on a background thread looking for results. It would be slower and less efficient, but it also seems less prone to complex synchronization issues.

Reply:
> Maybe we should change the overall approach

FWIW, I filed https://github.com/jenkinsci/workflow-support-plugin/pull/259 with an example of what that could look like.

Reply:
> Those threads will call `StepContext.onSuccess`/`StepContext.onFailure` before `StepExecution.start` has completed

That would be normal if `false` is returned. It would be weird if `start` later returns `true`, but I do not think it is harmful. TBH I do not recall the logic here in `workflow-cps` very well. Anyway this seems unlikely in a normal test, because it should call `wait` and then do some stuff before calling `success`.